JBPM-5735: Rule contexts, handlers & extensions.

romartin opened the pull request:
Hey @manstis , @hasys 
This commit is about some refactoring and improvements required on the rules stuff.This is basically a refactor of the current API, allowing generic rules and contexts, and the addition of an extensions mechanism as well. See the [Jira ticket](https://issues.jboss.org/browse/JBPM-5735).
Once this gets merged, I'll provide another PR for adding the BPMN emb. subprocess and its concrete rule extensions on top of this.
Thanks!
Roger

manstis commented:
Ahhh.. there's nothing quite like a @romartin PR to make your morning ;-)

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-client/src/test/java/org/kie/workbench/common/stunner/cm/client/canvas/controls/builder/CaseManagementElementBuilderControlTest.java:
```diff
@@ -22,7 +22,7 @@
 
 @RunWith(MockitoJUnitRunner.class)
-public class CaseManagementElementBuilderControlTest extends BaseCaseManagementBuilderControlTest {
+public class CaseManagementElementBuilderControlTest extends org.kie.workbench.common.stunner.cm.client.canvas.controls.builder.BaseCaseManagementBuilderControlTest {
```
Do we need FQCN?

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-client/src/test/java/org/kie/workbench/common/stunner/cm/client/canvas/controls/builder/CaseManagementObserverBuilderControlTest.java:
```diff
@@ -25,7 +25,7 @@
 
 @RunWith(MockitoJUnitRunner.class)
-public class CaseManagementObserverBuilderControlTest extends BaseCaseManagementBuilderControlTest {
+public class CaseManagementObserverBuilderControlTest extends org.kie.workbench.common.stunner.cm.client.canvas.controls.builder.BaseCaseManagementBuilderControlTest {
```
Do we need FQCN?

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/validation/graph/GraphValidatorImpl.java:
```diff
@@ -45,51 +48,42 @@
+    private final RuleManager ruleManager;
+    private RuleSet ruleSet;
 
     protected GraphValidatorImpl() {
```
Big important class with lots of changes without tests.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-processors/src/main/java/org/kie/workbench/common/stunner/core/processors/rule/ExtensionRuleGenerator.java:
```diff
@@ -0,0 +1,173 @@
+import org.uberfire.annotations.processors.AbstractGenerator;
+import org.uberfire.annotations.processors.exceptions.GenerationException;
+
+public class ExtensionRuleGenerator extends AbstractGenerator {
```
Not tested new class.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-processors/src/main/java/org/kie/workbench/common/stunner/core/processors/rule/ExtensionRuleGenerator.java:
```diff
@@ -0,0 +1,173 @@
+                                  value);
+    }
+
+    private StringBuffer generateRule(final Messager messager,
```
Not used parameter _messager_ in a private class.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-processors/src/main/java/org/kie/workbench/common/stunner/core/processors/rule/ExtensionRuleGenerator.java:
```diff
@@ -0,0 +1,173 @@
+        //Generate code
+        final StringWriter sw = new StringWriter();
+        final BufferedWriter bw = new BufferedWriter(sw);
+        try {
```
I guess [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) can be used here.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-processors/src/main/java/org/kie/workbench/common/stunner/core/processors/MainProcessor.java:
```diff
@@ -937,6 +957,38 @@ private boolean processContainmentRules(final Set<? extends TypeElement> set,
         return true;
     }
 
+    private boolean processRuleExtension(final Set<? extends TypeElement> set,
```
Big method with two unused parameters (_set_ and _roundEnv_), code duplication without tests.

P.S. looks like entire class MainProcessor has no tests at all, but is very important.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-core-api/src/main/java/org/kie/workbench/common/stunner/core/rule/annotation/RuleExtension.java:
```diff
@@ -0,0 +1,41 @@
+
+    Class<? extends RuleExtensionHandler> handler();
+
+    String[] arguments() default {};
```
well not exactly... arguments and typeArguments are not related, are just two arrays that can be used to provide arguments, one for raw strings and another if you need to use class types as arguments. Objects cannot be used on the anotations itself, so thsi is a try to achieve having serialized arguments where possible and having too class type argments, if the handler needs so.... I know it's not the ideal, but it's a try to make some annotation with "generic arguments"... what you think?

manstis replied:
@romartin OK fair enough.. ignore me!

romartin replied:
hehe np! thanks! ( will ignore you as I usually do.. :P )

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/rule/handler/impl/GraphEvaluationHandlerUtils.java:
```diff
@@ -48,12 +48,27 @@ protected String getElementDefinitionId(final Element<?> element) {
+                                                       final RuleViolation violation) {
+        try {
+            ((AbstractRuleViolation) violation).setUuid(uuid);
+        } catch (final ClassCastException ignored) {
```
Yes an exception can be expected, anyway probably will be better using `instanceof` rather than expecting an exception. 
It can be expected (this is new) because if the rule violation is done for some graph element, the `AbstractRuleViolation` has a public "uuid" that can be set, this way on the resulting violations you can obtain the uuid of the invalid elements and do whatever you need with the shapes to provide some feedback (as changing the state to invalid for example). Are you thinking on any other way to achieve this? :) thanks!

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/rule/handler/impl/ElementCardinalityEvaluationHandler.java:
```diff
@@ -0,0 +1,108 @@
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public RuleViolations evaluate(final Occurrences rule,
```
hey @hasys , this method is already tested by 'ElementCardinalityEvaluationHandlerTest', isnt it?

hasys replied:
Oh, yes, sorry my IDE lie to me :)

romartin commented:
Hey @manstis , @hasys 

Thanks both for the review! Added a new commit which fixes most of the comments above, just missing to add tests for the control accetpors and the annotation processors, which were already missing, but I think we can address those later on as we have a time frame for adding missing unit tests. Hope this works for you, I would like to have this merged as soon as possible so we can do bug fixing on top of it.
@hasys : I have reverted the subprocess to the previous shape, and we can address the right one as with [this ticket](https://issues.jboss.org/browse/JBPM-5550). 
@manstis  @hasys : For me it's ok creating a PR just for updating licenses yeah!

Thanks!
Roger

manstis commented:
@romartin Thanks.
@romartin @hasys Annotation Processor tests is covered [here](https://issues.jboss.org/browse/JBPM-5749).

romartin commented:
Hey @hasys @manstis , yeah I was not taking too much care on the `MainProcessor` and this related classes as I know they need a complete refactor and testing... anyway well seen @hasys , I could easily refactor that duplication code and remove the unused arguments. Hope this is ready to merge now!! :) Thanks guys!
